### DISCUSSION TITLE
The NewV2board send DNS config

### DISCUSSION THREAD

**Discussion by Septrum101** [XrayR-project/XrayR#154] [Ideas]
Hello @pocketW , V2board has sent this json to backend. I just add the block rules. So should we add DNS rules to the core instance?
```json
{
    "server_port": 10000,
    "network": "grpc",
    "networkSettings": {
        "path": "ray",
        "headers": {
            "Host": "test.com"
        }
    },
    "tls": 1,
    "base_config": {
        "push_interval": 60,
        "pull_interval": 60
    },
    "routes": [
        {
            "id": 1,
            "match": "360.cn",
            "action": "block",
            "action_value": null
        },
        {
            "id": 2,
            "match": "netflix.com",
            "action": "dns",
            "action_value": "1.1.1.1"
        }
    ]
}
```

I see the core instance have those interface
```
Type() interface{}
Close() error
RequireFeatures(callback interface{}) error
AddFeature(feature features.Feature) error
GetFeature(featureType interface{}) features.Feature
Start() error
```
Could it be use `GetFeature(featureType interface{})` to get exsit dns setting? And after add new DNS rules use `AddFeature(feature features.Feature)` to update DNS rules?

**Comment by pocketW** [comment]
Hi @thank243 , I think this feature is related to the routing management API in this PR XTLS/Xray-core/pull/412 . But I am not sure why this PR has not been merged. There is another issue discussing this features XTLS/Xray-core/issues/414. 

From my opinions, I think we can replace the [router](https://github.com/XTLS/Xray-core/tree/main/app/router) with our customized route package like we did for dispatch. It could allow us add a management API for XrayR to call and modify the route.

**Reply by Septrum101** [reply]
I test a demo. Direct write byte to dns.json. It works. So should we request   node config on package panel before start core.instance？

**Reply by pocketW** [reply]
It is not a good idea to directly modify dns.json. Since it can only be loaded at the start of the core. If the server updates the DNS info, we might need to restart the whole core service. This will break all the existing connections. If possible, it is better to modify the route item dynamically.

**Reply by pocketW** [reply]
Another possible solution is to get route features from core using [GetFeature](https://github.com/XTLS/Xray-core/blob/f35ded79ad3312eaa9054074e9f5448bae555664/core/xray.go#L333) and close it. Then init a new route features and add it back to the core using [AddFeature](https://github.com/XTLS/Xray-core/blob/f35ded79ad3312eaa9054074e9f5448bae555664/core/xray.go#L299)

**Reply by Septrum101** [reply]
I will check it a few days.

**Reply by Septrum101** [reply]
@pocketW I just test those method you mentioned. It return `*dns.DNS`, but all the struct's key are private. Thus we cannot replace it directly unless we implement all DNS interface. It'll take a lot of time. So I give it up. LOL
I think we should add a dnsConfig in controller and ervery period check compare the new and old dnsConfig, if changed, we can build a new DNSConfig{} directly replace whole dns in the core.instance?

**Reply by pocketW** [reply]
We can copy the [dns](https://github.com/XTLS/Xray-core/blob/main/app/dns/dns.go) package and make all the struct's key accessable. Then we can register this customized dns in [here](https://github.com/XrayR-project/XrayR/blob/master/main/distro/all/all.go) to replace the old one like we did for [dispatcher](https://github.com/XrayR-project/XrayR/tree/master/app/mydispatcher).

Another choice is to directly replace whole dns in core.instance. I think the `GetFeature` and `AddFeature` I mentioned above might help.

**Reply by Septrum101** [reply]
I'm considering whether the remote DNS should override the local. If override, the feat will be easy to implement, else I should add local DNS to controller.
